### PR TITLE
basic_gateway_design

### PR DESCRIPTION
**Pull request by zhangml**
This is the outline design of the basic gateway.

### CODE REVIEW

File: docs/design/basic_gateway/design.md [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+-   Only a small number of necessary rules are installed when the virtual machine goes online, and the remaining rules are gradually installed as needed after the virtual machine is online；
+
+-   The size of the rules installed when virtual machine is online is decoupled from the VPC scale；
+
+-   The number of rules deployed on the compute node is greatly reduced, reducing the risk of overload and failure of high-density servers；
+
+### 2. Gateway node forwarding model
+
+In order to reduce the number of rules installed when the virtual machine is online and increase the speed of the virtual machine going online, we designe a gateway node forwarding model. Each gateway node stores the global rule information of one VPC or multiple VPCs. In the gateway node forwarding model, when a virtual machine is online, the rules configuration is mainly divided into two parts:
```
Comment (zasherif) [inline docs/design/basic_gateway/design.md:15]: Are you suggesting that we rename bouncer to gateway. Why don't you simply configure group-select rules to send the packets to the bouncer; which already exist?

File: docs/design/basic_gateway/design.md [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+-   Installing the configuration related to the virtual machine to access the network, and configure the bucket-to-gateway mapping on the host-side OVS   (buckets are described below). At this point, the virtual machine has been successfully launched.
+
+-    When the newly-launched virtual machine communicates with other virtual machines, it needs to query the rules in the OVS flow table on the computing node. Since the newly-launched virtual machine only matches the default rules for forwarding to the gateway node, the first packet of the flow will be sent to the gateway node. Then, the gateway node queries the global rules and installs the direct path rules for the OVS in the compute node  where the virtual machine is located. Then the virtual machine pairs can use direct path forwarding if the rules have been installed.
```
Comment (zasherif) [inline docs/design/basic_gateway/design.md:19]: Are you going to query rules in the OVS flow table? How will you do that without buffering packets? Do you mean OVS matches existing rules?

Can you explain more what do you mean by: "the gateway node queries the global rules and installs the direct path rules for the OVS"?

File: docs/design/basic_gateway/design.md [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+![](png/gateway_model.png)
+
+<center>Figure 1. Gateway node model</center>
+
+### 3. Gateway node cluster management
```
Comment (zasherif) [inline docs/design/basic_gateway/design.md:25]: Without detailed information for my earlier questions on Section.2, this part is invalid.

File: docs/design/basic_gateway/design.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
+## On-demand installing of table entries based on basic gateways
```
Comment (zasherif) [inline docs/design/basic_gateway/design.md:1]: Is this "basic gateways" different from bouncers? If so can you detail their function? if not, can you just stick to the terminology?

**Review by zasherif — CHANGES_REQUESTED** [review]
This needs substantial iterations. We need a clear design of what is the role of the basic gateway. example of group-select rules, and how the "gateway node queries the global rules to do the direct-path"?

File: docs/design/basic_gateway/design.md (hunk 4 above)
Comment (zhangml) [reply docs/design/basic_gateway/design.md:1]: Okay, “basic gateways” and “bouncers” are really different.  We will update more details in later documents.

Comment (zasherif) [reply docs/design/basic_gateway/design.md:1]: Sure thanks. would you please make sure to address all questions in the next revision?

**Comment by vinaykul** [comment]
This PR has become stale. It will most likely be superceded by Fornax project implementations. Closing.

**Pull request closed without merging**